src/Microsoft.SqlTools.ServiceLayer/Agent/Jobs/AgentNotebookHelper.cs

```diff
@@ -620,26 +635,75 @@ UPDATE notebooks.nb_materialized
 
         public static async Task UpdateMaterializedNotebookPin(
             ConnectionInfo connInfo,
-            string materializedId,
+            AgentNotebookHistoryInfo agentNotebookHistory,
             string targetDatabase,
             bool pin)
         {
             string updateMaterializedNotebookPinQuery =
-             @"
+            @"
+            IF EXISTS
+            (SELECT * FROM notebooks.nb_materialized 
+            WHERE job_id = @jobId AND run_time = @startTime AND run_date = @startDate)
+            BEGIN
                 UPDATE notebooks.nb_materialized 
                 SET 
-                pin = @notebookPin 
+                pin = @notebookPin
                 WHERE 
-                materialized_id = @materializedId
+                job_id = @jobId AND run_time = @startTime AND run_date = @startDate
+            END
+            ELSE
+            BEGIN
+                INSERT INTO notebooks.nb_materialized (job_id, run_time, run_date, notebook, notebook_error, pin) 
+                VALUES 
+                (@jobID, @startTime, @startDate, '', '', @notebookPin)
+            END
             ";
             List<SqlParameter> updateMaterializedNotebookPinParams = new List<SqlParameter>();
+            updateMaterializedNotebookPinParams.Add(new SqlParameter("jobID", agentNotebookHistory.JobId));
+            updateMaterializedNotebookPinParams.Add(new SqlParameter("startTime", agentNotebookHistory.RunDate.ToString("HHmmss")));
+            updateMaterializedNotebookPinParams.Add(new SqlParameter("startDate", agentNotebookHistory.RunDate.ToString("yyyyMMdd")));
             updateMaterializedNotebookPinParams.Add(new SqlParameter("notebookPin", pin));
-            updateMaterializedNotebookPinParams.Add(new SqlParameter("materializedId", materializedId));
             await AgentNotebookHelper.ExecuteSqlQueries(
                 connInfo,
                 updateMaterializedNotebookPinQuery,
                 updateMaterializedNotebookPinParams,
                 targetDatabase);
         }
+
+        public static async Task DeleteMaterializedNotebook(
+            ConnectionInfo connInfo,
+            AgentNotebookHistoryInfo agentNotebookHistory,
+            string targetDatabase)
+        {
+            string deleteMaterializedNotebookQuery =
+            @"
+            IF EXISTS
+            (SELECT * FROM notebooks.nb_materialized 
+            WHERE job_id = @jobId AND run_time = @startTime AND run_date = @startDate)
+            BEGIN
+                UPDATE notebooks.nb_materialized 
+                SET is_deleted = 1,
```

Comment:
so we don't want to actually delete the rows?  I think either design is fine, though, my first thought was that deleted would actually remove the data.